airflow/bin/cli.py

```diff
@@ -70,10 +70,18 @@
 
 LOG = LoggingMixin().log
 
-DAGS_FOLDER = settings.DAGS_FOLDER
 
-if "BUILDING_AIRFLOW_DOCS" in os.environ:
-    DAGS_FOLDER = '[AIRFLOW_HOME]/dags'
+def get_dags_folder():
+    """Retrieves DAGS_FOLDER to use."""
+    from airflow.settings import DAGS_FOLDER
```

Comment:
I feel that since this fn is called at the module level this code is identical to what was there before.

Reply:
It's not the same actually. It's quite different because the scope is local, so you cannot be lazy and import all settings like that without knowing the consequences.

We have now automated check in pre-commit which checks for '^from airflow.settings import'. Which means that you will not be able to import some settings directly at the top level of your python file. Which is a good thing knowing that importing settings is kind of poisonous - and can easily lead to cyclic imports. You need to make a deliberate effort to import some variable now and you will think twice before you do it in and pre-commit fails.

What I want to do here is to promote the pattern that all setting imports are done locally. Usually (psychology) where people will try to import the settings directly at the top level the pre-commit fail and even if they do not use pre-commit they will (naturally) look on how settings are used elsewhere. And they will find that almost universally airflow.settings are imported individually and locally.

There are only three exception where it makes sense and I allow to import settings. And they are all legitimate places - and they should not be imported anywhere else.

```
          ^airflow/__init__.py$|airflow/plugins_manager.py|
          airflow/www/gunicorn_config.py
```

Reply:
> and can easily lead to cyclic imports

If it would lead to an import cycle here then importing at the top level, or importing at a function which is called from the top level (i.e. this) would be no differnce - that was what I meant by "same"

My main issue I think with this PR is that changing from `from airflow import settings` to `from airflow.settings import SOME_VAR` does nothing to change or prevent import cycles. If there was an import cycle under one so would there be under the other.

And given we _already_ have tests (mypy) that will fail for import cycles why is any of this necessary?

Reply:
In this particular case - yes it does not change the "cyclability" but it promotes some developer behaviours (i.e. using local imports with single fields rather than import airflow.settings).

But you are quite right that in this particular case (cli.py) it is perfectly ok to load settings at the top level because it is yet another "entry-point". 
That's why I reverted this particular change and excluded bin/cli.py from the pre-commit check.

For all the other cases it is super important that we are promoting good behaviours. 

What we are flagging now is:

- from airflow import.* settings -> importing settings as a whole (whether global or local)
- import airflow.settings -> also importing settings as a whole (whether global or local)
- ^from airflow.settings --> importing any of the settings objects at top level (see ^) at the beginning

This means that what is actually good is importing particular settings as local imports. And all the examples of using settings are like that in the code, so I assume people when adding settings and failing on pre-commit will see how settings are used and copy that behaviour. I can also add some extra explanation in CONTRIBUTING.rst but I thought that we check if people have questions - we can always add it later or add more meaningful description in case of detection.

Forcing local imports for settings changes everything when it comes to cycles. It means that settings (and it's transitive dependencies) will not be randomly firing. 

The problem we are trying to solve is that those cycles are not detected always. This partially due to the way pylint works (where it starts to check for imports). We have so complex that It is not deterministic - when you add new files, you sometimes end up with cycles and sometimes you don't - depending on where pylint starts. It happened already a number of times for different people that the cycle was detected after they added a completely irrelevant file. This is also likely due to the way pylint checks are done in Travis/pre-commit. it is run in parallel (2 threads) on roughly half of the files so it might well be that it is not properly adding all the dependent files while running the check.

Reply:
Also just to explain: @ashb . I realised you might not be aware so I want to give you the context - all this effort is really needed. It's not an artificial problem. it's a real one that we have to fix and prevent from happening in the future. 

Here is the story: last week several people (@ratb3rt can confirm) started to have seemingly random pylint failures in their PRs caused by cyclic imports (in files the people were not changing !)

You can see for example here, where totally unrelated change cause cyclic import reported by pylint: 
https://travis-ci.org/apache/airflow/jobs/612096100?utm_medium=notification&utm_source=github_status

Same kind of problems were reported by @TobKed in some of his builds.

It also happened in master after I merged a change which was OK before merge 

https://travis-ci.org/apache/airflow/jobs/613266680?utm_medium=notification&utm_source=github_status

And in CRON build:
https://travis-ci.org/apache/airflow/jobs/613319288?utm_medium=notification&utm_source=github_status

And it's been fixed by subsequent merge of unrelated changes (!)

Here is the link to slack discussion where I discussed it with @ratb3rt : https://apache-airflow.slack.com/archives/CCPRP7943/p1573589242010400

I investigated it as a weakness of pylint detecting the cycles only sometimes - depending on the sequence/set of files pylint will get as input. Seems that depending on what is the sequence/set of files passed to pylint, it will either not detect the cycle (most of the times) or it will detect it (quite rare but can happen). 

The main thing - the cycle really IS there, it's just masked - depending on which file gets imported first.

So what I want to do in this and subsequent changes is to:

a) remove the cycles
b) add some protection to make it more difficult to introduce such cycles accidentally in the future.

I really hope we can solve it permanently.

Reply:
Pylint saying there is an import cycle is very different to there being an actual import cycle that will break running Python and cause a recursion error.

Reply:
What do you propose @ashb ? Should we drop pylint? Or fix it somehow so that it does not show cycles ? Because if we do nothing, then people will continue getting random pylint errors.

Reply:
That's [true](https://github.com/PyCQA/pylint/issues/850). 

> Settings loads a lot of dependencies on its own and having it imported at the top of the file often causes many transitive/implicit dependencies which might cause circular imports.

@potiuk isn't that then an issue with the settings module itself instead of the ones that are importing it? I would suggest that we split up those bigger modules into smaller ones.

Reply:
I think if we introduce these changes here we are working around the bug but not entirely fixing it. I think most of these errors occur because of bad modules (structure).

Reply:
@ashb @feluelle . As mentioned in slack I went back to the drawing board and looked more closely into the problems and I would like to introduce a series of commits that will solve it better than I initially thought. Indeed local import in settings kind of masked the problem, so I went back to see what was the real problem if we leave it as it was.

I think the main problem was really the CONTEXT_MANAGER_DAG in settings and especially the type annotation there for the DAG class. This was the main problem with settings.

So I left in this commit/PR  and changed JIRA to reflect it is only to move the DAG context manager to DAG. It's much smaller and less disruptive and would not require to change any of the user's patterns (such as importing settings). This is the first step, more will follow.